### PR TITLE
fix: allow kubeflow api endpoint to have scheme as optional

### PR DESCRIPTION
**Pull request by rkpattnaik780**
### What changes were proposed in this pull request?
1. Modify logic so that API endpoint shoiuld have the URL scheme as an optional.

### How was this pull request tested?
1. Run `elyra-metadata list runtimes`.
2. Modify the api endpoint.
3. You should see that Elyra UI's runtime section loads the runtime without issue.

Fixes [RHOAIENG-5071](https://issues.redhat.com/browse/RHOAIENG-5071)
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: elyra/metadata/schemas/kfp.json [hunk 1]
```diff
@@ -47,7 +47,7 @@
           "title": "Kubeflow Pipelines API Endpoint",
           "description": "The Kubeflow Pipelines API endpoint",
           "type": "string",
-          "format": "uri",
+          "pattern": "(?:(?:http|https)://)?(www\\.)?[a-zA-Z0-9@:%._\\+~#?&//=]{2,256}\\.[a-z]{2,6}\\b([-a-zA-Z0-9@:%._\\+~#?&//=]*)",
```
Comment (rkpattnaik780) [inline elyra/metadata/schemas/kfp.json:50]: While this allows schema less URLs to be accepted, no error is thrown when the input doesn't match the regexp.

Comment (jiridanek) [reply elyra/metadata/schemas/kfp.json:50]: Wow, that's quite an regex. What about stealing some ideas from https://www.stephenlewis.me/blog/json-schema-url-validation/? They rely on format/hostname in json schema spec, https://json-schema.org/understanding-json-schema/reference/string#hostnames, plus the `anyOf` operator.

When I leave protocol out, does that imply `https`, or is it `http or https`?

Can this be tested here? https://github.com/opendatahub-io/elyra/blob/3f11b019939effbac016862f8caffda8f560e724/elyra/tests/cli/resources/runtime_configs/valid_kfp_test_config.json#L4

Comment (rkpattnaik780) [reply elyra/metadata/schemas/kfp.json:50]: Thanks for sharing this @jiridanek . While this is indeed a much cleaner way to perform the validation. It am not able to get through scenario of `.com/v1` or `.com/`. Let me check if I can modify the pattern to handle these cases as well.

Comment (jiridanek) [reply elyra/metadata/schemas/kfp.json]: > It am not able to get through scenario of .com/v1 or .com/.

The article actually said at the end that with their approach this does not work. I missed it when I was reading it ;(

> ## Scheme-less URLs
>
> But what about “scheme-less” URLs, such as www.example.com?
>
> This is where JSON Schema comes up short. If you’re dealing with just a domain, and no path, the following definition covers the basics. It fails miserably on something like www.example.com/endpoint, though.
>
> ...
>
> Either way, you’ll still need to do some manual validation.

File: elyra/metadata/schemas/kfp.json [hunk 2]
```diff
@@ -47,7 +47,7 @@
           "title": "Kubeflow Pipelines API Endpoint",
           "description": "The Kubeflow Pipelines API endpoint",
           "type": "string",
-          "pattern": "(?:(?:http|https)://)?(www\\.)?[a-zA-Z0-9@:%._\\+~#?&//=]{2,256}\\.[a-z]{2,6}\\b([-a-zA-Z0-9@:%._\\+~#?&//=]*)",
+          "pattern": "^((?:https?:\\/\\/)?[^./]+(?:\\.[^./]+)+(?:\\/.*)?)$",
```
Comment (jiridanek) [inline elyra/metadata/schemas/kfp.json:50]: yes, this looks much better than the previous regex!

IIUC, looking at the structure of error message in https://issues.redhat.com/browse/RHOAIENG-5071, the validation is performed at

https://github.com/opendatahub-io/elyra/blob/3f11b019939effbac016862f8caffda8f560e724/elyra/metadata/manager.py#L159

File: elyra/metadata/schemas/kfp.json [hunk 3]
```diff
@@ -47,7 +47,7 @@
           "title": "Kubeflow Pipelines API Endpoint",
           "description": "The Kubeflow Pipelines API endpoint",
           "type": "string",
-          "format": "uri",
+          "pattern": "^((?:https?:\\/\\/)?[^./]+(?:\\.[^./]+)+(?:\\/.*)?)$",
```
Comment (jiridanek) [inline elyra/metadata/schemas/kfp.json:50]: Security: How do we know that when I leave out https://, Elyra will still use https (and not http)?

Comment (rkpattnaik780) [reply elyra/metadata/schemas/kfp.json:50]: I have modified the logic to append "https" if the scheme is missing as secured connection is now mandated.

Comment (jiridanek) [reply elyra/metadata/schemas/kfp.json:50]: lgtm, I guess

File: elyra/metadata/schemas/kfp.json (hunk 2 above)
Comment (harshad16) [reply elyra/metadata/schemas/kfp.json:50]: This regexp is great, though we dont need it for the kubeflow pipeline endpoint only for the cos_endpoint that is below, can you revert this line back please.

**Comment by rkpattnaik780** [comment]
@harshad16 , with this PR, the https gets added to the generated yaml and python file if it is missing in the metadata json .

From the python file generated by export:
```
--pipeline-name 'hello-generic-world' --cos-endpoint 'https://minio-minio.apps.rosa.rpattnai-rosa.echt.p3.openshiftapps.com' --cos-bucket
```

From the pipeline spec:
Screenshot 2024-05-20 at 5 07 38 PM

Elyra UI:
Screenshot 2024-05-20 at 5 06 01 PM

**Review by harshad16 — APPROVED** [review]
/lgtm
/approve

thanks :100:

**Pull request merged**